### ISSUE TITLE
Algorithm explanation

### ISSUE THREAD

**Issue by ghost** [dfaker/df#20]
Hi dfaker, if you don't mind, can I ask some questions about the implementation? I'm not sure where else to ask other than here.

For the larger faces, I see that you pull out 240x240 pixels from the 256x256 cropped faces. The original code uses 160x160, like at the faceswap repo.

Is anything else in your code specific to generating larger face models? The model is more complex, but I think that most of it is related to generating a higher resolution 128x128 output, not for generating larger face models?

In other words, if I took the faceswap repo, and just change the image_augmentation to mimic yours with 240x240, would that generate large face models at 64x64 resolution? Or is there some other step that needs to be changed to make it work? Are changes in the conversion process essential to the large face model, or wouldn't just using user-defined kernel/blur etc. be sufficient?

**Comment by dfaker** [comment]
The changes in image_augmentation are more to address negatives in the original warping process, that it didn't target the areas of the face that vary and didn't really focus on transformations specifically useful in the transfer task, the old method still gets reasonable results.

The penalized_loss in model.py is pretty vital as the image is zoomed out, it prevents the model from trying to learn non-facial features, otherwise the random background elements can saturate the limited memory available for storing a description of the face.

**Comment by ghost** [comment]
Thanks! I wish I could "git clone dfaker's brain". Learning a lot going through your code. Will leave this open for now in case I have other questions, so I don't spawn too many issues.

**Comment by ghost** [comment]
Alright, I'll try here. I know you mentioned halving the resolution in jest for  a reddit reply ;), but I was actually trying to do that for real in another place on the faceswap repo. Do you know why just changing IMAGE_SHAPE=(32,32,3) and cv2.resize on the warped and target image to (32,32) doesn't work? Or maybe setting the Encoder output to 32 as well? This is the naive approach - why does it fail?

Like in dfaker, there seem to be 3 improvements:
-larger face crop
-algorithmic tweaks to warping, masking, etc.
-higher output resolution

I'm trying to figure out the dead simple change of just lower (or higher) input/output resolution. It's a bit convoluted in dfaker because of the other 2 improvements. I assume input was set at 64 because of memory/cost issues. But stepping both the input/output 32x32 - shouldn't that be straightforward, like in the naive approach above?

**Comment by dfaker** [comment]
You'll want to use `print(model.summary())` on the autoencoders in model.py to check their input and output size, basically each of the `x = upscale(256)(x)` layers is doubling each dimension in the decoders and each `x = conv( 256)(x)` is halving it in the encoder.

**Comment by ghost** [comment]
Thank you!
